File: 프로그래머스/lv3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;

class Solution {
    // 하드 디스크를 가장 효율적이게 순서대로 처리하는 방법?
    // 끝나는 시점이 빠른 것부터 순차적으로 돌리자
    
    // 구해야하는 값 = 소수점을 제외한 작업 요청부터 걸린 시간까지의 평균
    
    // 중요 : 하드디스크가 쉬는 동안에는 무조건 시작이 빠른 게 장땡임
    public int solution(int[][] jobs) {
        Arrays.sort(jobs, (o1, o2) -> o1[0] - o2[0]);
        Queue<int[]> queue = new PriorityQueue<>((o1, o2) -> o1[1] - o2[1]);
        
        int answer = 0;
        int count = 0;
        int idx = 0;
        int end = 0;
        
        while (count < jobs.length) {
            while (idx < jobs.length && jobs[idx][0] <= end) {
                queue.add(jobs[idx++]);
            }
            
            if (queue.isEmpty()) {
                end = jobs[idx][0];
            } else {
                int[] job = queue.poll();
                answer += job[1] + end - job[0];
                end += job[1];
                count++;
            }
        }
        
        return answer/jobs.length;
    }
}

File: 프로그래머스/lv3/42627. 디스크 컨트롤러/디스크 컨트롤러.kt
import java.util.*

class Solution {
    fun solution(jobs: Array<IntArray>): Int {
        val orderedJobs: MutableList<IntArray> = jobs
            .toList()
            .sortedWith(compareBy ({ it[0] }, { -it[1] }))
            .toMutableList()
            
        val delayTimePriorityQueue: PriorityQueue<IntArray> = PriorityQueue() {
            a, b -> 
            if (a[1] != b[1]) {
                a[1] - b[1]
            } else a[0] - b[0]
        }
            
        var idx: Int = 0
        var time: Int = 0
        var endTime: Int = 0
        var count: Int = 0
        
        while (count < orderedJobs.size) {
            while (idx < orderedJobs.size && orderedJobs[idx][0] <= endTime) {
                delayTimePriorityQueue.add(orderedJobs[idx++])
            }
            
            if (delayTimePriorityQueue.isEmpty()) {
                endTime = orderedJobs[idx][0]
            }
            
            if (!delayTimePriorityQueue.isEmpty()) {
                val currentJob: IntArray = delayTimePriorityQueue.poll()
                println(currentJob.toList().toString())
                
                time += (endTime - currentJob[0]) + currentJob[1]
                endTime += currentJob[1]
                println("endTime : $endTime currentJob[0] : $currentJob[0] currentJob[1] : $currentJob[1] => time = $time")
                count++
            }
        }
        
        return time / jobs.size
    }
}